Big Keymap Restructure

algernon opened the pull request:
This changes the keymap layout to be as described in #17, and updates the default layout to follow. (Also adds the missing Space and Mod keys on the two keys below the bottom row)

The layout itself was considerably simplified by introducing a few local macros to hide some of the uglier details.

Also includes #16, rebased and slightly redone, on top of the restructuring.

Quick tests show that at least the `MOD` layer is working, as does the base.

algernon committed:
Big Keymap Refactor(tm)

This changes the keymap layout to be as described in #17, and updates the
default layout to follow. (Also adds the missing Space and Mod keys on the two
keys below the bottom row)

The layout itself was considerably simplified by introducing a few local macros
to hide some of the uglier details.

Fixes #17.

Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

algernon committed:
Better key state tracking

We need to store the previous state of the keys, to figure out when a specific
key is released, as opposed to simply not being pressed. This is required for
layer keys such as `Mod` and `Fn` to work properly when there are more than one
of them on the keymap.

Without being able to tell when a key is released, we would not be able to turn
a layer off only when no layer key is held. Not easily, anyway: we'd have to
track the state of all layer keys... but then it is easier to just track them
all.

Thankfully, the memcpy at the end is fast, as it only needs to copy 70 bytes.

Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

mondalaci commented on right/src/default_layout.c:
```diff
@@ -1,77 +1,56 @@
-        {{.type = UHK_KEY_SIMPLE, .simple = { .key = HID_KEYBOARD_SC_RIGHT_GUI } }},
-        {{.type = UHK_KEY_SIMPLE, .simple = { .key = HID_KEYBOARD_SC_RIGHT_CONTROL } }},
+#define S(k) {.type = UHK_KEY_SIMPLE, .simple = { .key = HID_KEYBOARD_SC_ ## k }}
+#define L(l) {.type = UHK_KEY_LAYER, .layer = { .target = LAYER_ID_ ## l }}
```
Would you please rather not use these macros? I understand that they make the code shorter but it'd be great to be able to see the keymap structure as it is. The reusability of the macros shouldn't be a significant benefit either  because we will only define the factory keymap in the firmware. All the other keymaps will live in the EEPROM, serialized by Agent.

algernon replied:
Sure thing. They were mostly there to help me while translating the structure to the new format. Will expand them.

mondalaci commented on right/src/default_layout.c:
```diff
@@ -1,77 +1,56 @@
-        {{.type = UHK_KEY_SIMPLE, .simple = { .key = HID_KEYBOARD_SC_RIGHT_CONTROL } }},
+#define S(k) {.type = UHK_KEY_SIMPLE, .simple = { .key = HID_KEYBOARD_SC_ ## k }}
+#define L(l) {.type = UHK_KEY_LAYER, .layer = { .target = LAYER_ID_ ## l }}
+#define TRNS {.type = UHK_KEY_TRANSPARENT}
```
Would you please not use this type? It isn't one of the key actions supported by Agent.

algernon replied:
It's a hack, kind of, to make it possible to not repeat the layer switchers on the target layer too. But I guess I can just repeat it, doesn't matter.

mondalaci commented on right/src/default_layout.c:
```diff
@@ -1,77 +1,56 @@
+#define L(l) {.type = UHK_KEY_LAYER, .layer = { .target = LAYER_ID_ ## l }}
+#define TRNS {.type = UHK_KEY_TRANSPARENT}
+
+#define Key_NoKey { .raw = 0 }
```
I can't see the purpose of the `raw` member. How about adding a none action to `uhk_key_t` based on our official type and removing `raw`? The outer union type of `uhk_key_t` could be removed, leaving only the inner struct.

algernon replied:
That works too, yes. `.raw` comes from an earlier idea, when I was trying to push everything within 16 bits, and would have used it for some funky tricks. It is, as you correctly spotted, unnecessary by now.

Will drop it, thanks for the note!

mondalaci commented on right/src/default_layout.c:
```diff
@@ -1,77 +1,56 @@
+
+#define Key_NoKey { .raw = 0 }
+
+KEYBOARD_LAYOUT(defaultKeyboardLayout)= {
```
`CurrentKeymap` would be a better name because we'll always overwrite this keymap with the current one. Its initial content just happens to be the factory keymap.

It'd also be worth not to use a macro here. The reason is because we won't define another keymap anywhere, so it's not like we'll reuse the macro.

algernon replied:
ACK, drop macro, rename variable.

mondalaci commented on right/src/default_layout.c:
```diff
@@ -1,77 +1,56 @@
+    {
+        // Right
+        {
+            S(7_AND_AMPERSAND), S(8_AND_ASTERISK), S(9_AND_OPENING_PARENTHESIS), S(0_AND_CLOSING_PARENTHESIS), S(MINUS_AND_UNDERSCORE), S(EQUAL_AND_PLUS), S(BACKSPACE),
```
1 key action per line would read very nicely, especially when expanding key actions.

mondalaci commented on right/src/keyboard_layout.c:
```diff
@@ -1,22 +1,25 @@
 static uint8_t modifierState = 0;
 
-static inline __attribute__((always_inline)) uhk_key_t getKeycode(KEYBOARD_LAYOUT(layout), uint8_t keyId)
+static uint8_t prevLeftKeyStates[KEY_STATE_COUNT];
```
Ideally, we shouldn't have different variables on a per module basis but store both the left and right previous key states in an array of modules, similar to the keymap.

algernon committed:
Remove some now obsolete defines

Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

algernon committed:
Changes based on the review notes

This does a lot of things - all of this together, because it would have been
much harder to split them up into many small commits.

We get rid of the helper macros used in default_layout.c, drop the .raw member
of uhk_key_t, use a singleton keymap, and prepare the prevKeyStates to handle
all slots. Also drops the TRNS macro.

With these changes, default_layout.h became obsolete, and was deleted too.

Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

algernon committed:
Coding style fixes

Don't put a space between the function and the opening parens.

Signed-off-by: Gergely Nagy <algernon@madhouse-project.org>

mondalaci commented:
All good, thank you for the great work!
